src/jit/valuenum.cpp

```diff
@@ -8202,17 +8313,20 @@ void Compiler::fgValueNumberTree(GenTree* tree)
             case GT_HW_INTRINSIC_CHK:
 #endif // FEATURE_HW_INTRINSICS
             {
-                // A bounds check node has no value, but may throw exceptions.
-                ValueNumPair excSet = vnStore->VNPExcSetSingleton(
-                    vnStore->VNPairForFunc(TYP_REF, VNF_IndexOutOfRangeExc,
-                                           vnStore->VNPNormalPair(tree->AsBoundsChk()->gtIndex->gtVNPair),
-                                           vnStore->VNPNormalPair(tree->AsBoundsChk()->gtArrLen->gtVNPair)));
-                excSet =
-                    vnStore->VNPExcSetUnion(excSet, vnStore->VNPExceptionSet(tree->AsBoundsChk()->gtIndex->gtVNPair));
-                excSet =
-                    vnStore->VNPExcSetUnion(excSet, vnStore->VNPExceptionSet(tree->AsBoundsChk()->gtArrLen->gtVNPair));
+                ValueNumPair vnpIndex  = tree->AsBoundsChk()->gtIndex->gtVNPair;
+                ValueNumPair vnpArrLen = tree->AsBoundsChk()->gtArrLen->gtVNPair;
+
+                // Construct the exception set for bounds check
+                ValueNumPair vnpExcSet = vnStore->VNPExcSetSingleton(
+                    vnStore->VNPairForFunc(TYP_REF, VNF_IndexOutOfRangeExc, vnStore->VNPNormalPair(vnpIndex),
```

Comment:
Beware that `GT_HW_INTRINSIC_CHK` may throw exceptions other than `IndexOutOfRange`, `ArgumentOutOfRange` AFAIR. Not sure if it really matters for most purposes.

Reply:
Hmm, this seems to happen for `GT_ARR_BOUNDS_CHECK` as well, see simd.cpp around lines 2640-2650.

Reply:
SIMD nodes currently don't really participate with Value Numbering as they always get unique value numbers.  I have some ideas for future work that will allows us to remove more redundant exception checks based on Value Numbers.

Reply:
Well, they do not participate in value numbering but they still generate exceptions that VN records. What's not very clear to me if it makes any difference if we record IndexOutOfRange or ArgumentOutOfRange